core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,150 @@
+package ca.jrvs.apps.grep;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.io.*;
+import java.util.*;
+import java.util.regex.*;
+import org.apache.log4j.BasicConfigurator;
+
+public class JavaGrepImp implements JavaGrep {
+
+  final Logger logger = LoggerFactory.getLogger(JavaGrep.class);
+
+    private String regex;
+    private String rootPath;
+    private String outFile;
+
+
+
+  @Override
+  public void process() throws IOException {
+    List<String> matchedLines = new ArrayList<>();
+    for (File file : listFiles(getRootPath())) {
+      for (String line : readLines(file)) {
+        if (containsPattern(line)) {
+          matchedLines.add(line);
+        }
+      }
+    }
+    writeToFile(matchedLines);
+
+  }
+
+  @Override
+  public List<File> listFiles(String rootDir) {
+    List<File> listFiles = new ArrayList<>();
+    File root = new File(rootDir);
+    if (root.listFiles() != null) {
+      listFilesRecursively(root, listFiles);
+    }
+    return listFiles;
+  }
+
+  public void listFilesRecursively(File root, List<File> listFiles) {
+    File[] listOfFiles = root.listFiles();
+    if(listOfFiles == null) {
+      throw new NullPointerException("ERROR: root contains no files");
+    }
+    for (File file : listOfFiles) {
+        if (file.isDirectory()) {
+          listFilesRecursively(file, listFiles);
+        }
+        else {
+          listFiles.add(file);
+        }
+    }
+  }
+
+  @Override
+  public List<String> readLines(File inputFile){
+    if(!inputFile.isFile()) {
+      throw new IllegalArgumentException("ERROR: inputFile is not a file");
+    }
+    List<String> lines = new ArrayList<>();
+
+    // both BufferedReader and FileReader within try to catch both exceptions
+    try(BufferedReader bufferedReader = new BufferedReader(new FileReader(inputFile))) {
+      String line;
+      while((line = bufferedReader.readLine()) != null) {
+        lines.add(line);
+      }
+    } catch (Exception ex) {
+      logger.error("ERROR: BufferedReader or FileReader failed", ex);
+    }
+    return lines;
+  }
+
+  @Override
+  public boolean containsPattern(String line) {
+    Pattern pattern = Pattern.compile(getRegex());
+    Matcher matcher = pattern.matcher(line);
+
+    return matcher.matches();
+  }
+
+  @Override
+  public void writeToFile(List<String> lines) throws IOException {
+    File outFile = new File(getOutFile());
+    FileOutputStream fileOutStream = new FileOutputStream(outFile);
+
+    try(BufferedWriter bufferedWriter = new BufferedWriter(new OutputStreamWriter(fileOutStream))) {
+      for (String line : lines){
+        bufferedWriter.write(line);
+        bufferedWriter.newLine();
+      }
+    } catch (Exception ex) {
+      logger.error("ERROR: Write to outFile failed", ex);
```

Comment:
line 87 throws IOException, so you might not need to catch it. 
If you do need to catch the exception, make sure you throw a runtime exception. Otherwise, the program won't stop ..